ohttp/src/lib.rs

```diff
@@ -259,34 +275,35 @@ impl ClientRequest {
     #[allow(clippy::similar_names)] // for `sk_s` and `pk_s`
     pub fn new(encoded_config: &[u8]) -> Res<Self> {
         let mut config = KeyConfig::parse(encoded_config)?;
+        // TODO(mt) choose the best config, not just the first.
+        let selected = config.select(config.symmetric[0])?;
 
-        let hpke = HpkeS::new(
-            // TODO(mt) choose the best config, not just the first.
-            config.select(config.symmetric[0])?,
-            &mut config.pk,
-            INFO_REQUEST,
-        )?;
+        // Build the info, which contains the message header.
+        let info = build_info(config.key_id, selected)?;
+        let hpke = HpkeS::new(selected, &mut config.pk, &info)?;
 
-        Ok(Self {
-            key_id: config.key_id,
-            hpke,
-        })
+        let header = Vec::from(&info[INFO_REQUEST.len() + 1..]);
+        debug_assert_eq!(header.len(), REQUEST_HEADER_LEN);
+        Ok(Self { hpke, header })
     }
 
     /// Encapsulate a request.  This consumes this object.
     /// This produces a response handler and the bytes of an encapsulated request.
     pub fn encapsulate(mut self, request: &[u8]) -> Res<(Vec<u8>, ClientResponse)> {
-        // AAD is this header
-        let mut enc_request = Vec::new();
-        write_uint(size_of::<KeyId>(), self.key_id, &mut enc_request)?;
-        write_uint(2, u16::from(self.hpke.kem()), &mut enc_request)?;
-        write_uint(2, u16::from(self.hpke.kdf()), &mut enc_request)?;
-        write_uint(2, u16::from(self.hpke.aead()), &mut enc_request)?;
-
-        let mut ct = self.hpke.seal(&enc_request, request)?;
+        let extra =
+            self.hpke.config().kem().n_enc() + self.hpke.config().aead().n_m() + request.len();
+        let expected_len = self.header.len() + extra;
+
+        let mut enc_request = self.header;
+        enc_request.reserve_exact(extra);
```

Comment:
This bit is pure optimization on my part (no doubt premature), but I felt like I had to compensate for the new allocation I introduced.